### PR TITLE
Update attention layers with simpler flat forms, update wmt example.

### PR DESCRIPTION
**Pull request by levskaya**
refactoring wmt example in progress.

### CODE REVIEW

File: flax/linen/attention.py [hunk 1]
```diff
@@ -363,14 +250,13 @@ def __call__(self,
...
         dropout_rng=dropout_rng,
         dropout_rate=self.dropout_rate,
         broadcast_dropout=self.broadcast_dropout,
-        deterministic=self.deterministic)
+        deterministic=self.deterministic,
+        dtype=self.dtype,
+        precision=self.precision)
 
     # back to the original inputs dimensions
     out = DenseGeneral(features=features,
```
Comment (MostafaDehghani) [inline flax/linen/attention.py:262]: now that things are efficient, clean and simple, don't we want to also get rid of  `DenseGeneral`s as well?

**Review by MostafaDehghani — CHANGES_REQUESTED** [review]

File: flax/linen/attention.py [hunk 2]
```diff
@@ -33,147 +33,99 @@
...
-                          broadcast_dropout=True,
-                          dropout_rng=None,
-                          dropout_rate=0.,
-                          deterministic=False,
-                          precision=None):
+def dot_product_attention(query: Array,
+                          key: Array,
+                          value: Array,
+                          bias: Optional[Array] = None,
+                          broadcast_dropout: bool = True,
```
Comment (jheek) [inline flax/linen/attention.py:44]: Do you think we should get rid of the "inlined dropout" too?
One way is to make the attention function dynamic like in the design tests.

**Review by jheek — COMMENTED** [review]
Looks good to me! I have one small question

File: flax/linen/attention.py (hunk 1 above)
Comment (levskaya) [reply flax/linen/attention.py:262]: Hmm, I think the use of DenseGeneral here and above is nice, actually.  What you often do otherwise is define two "split_heads", "join_heads" util functions that do the two reshapes... and DenseGeneral is sort of just a nice concise way to do that.  Since we used it in mlperf I also know that it performs fine.

File: flax/linen/attention.py (hunk 2 above)
Comment (levskaya) [reply flax/linen/attention.py:44]: I agree that we should decouple it more.  it's already a dynamic attribute but the expected signature of the Callable is too specialized, it should just take q,k,v,bias.  I think I'd rather do that change in a separate PR though.

File: flax/linen/attention.py (hunk 1 above)
Comment (MostafaDehghani) [reply flax/linen/attention.py:262]: Totally agree that `DenseGeneral` is quite nice.  I was just referring to a tiny bit of improvement we can get from removing it, which might not be true given the extra reshape/transpose stuff we add with "split_heads", "join_heads" (which also makes the code a bit ugly).

p.s.
Just for the record, I profiled these two cases (which is wat is happening in DenseGeneral as well, the flattening applying dot_general on on axis was slightly better than applying dot_general on multiple axis:

```pyhton
rng = random.PRNGKey(0)
rng1, rng2 = random.split(rng)

inputs_shape = (8, 32, 32, 16, 16)  # bs, h, w, heads, features
inputs = random.normal(rng1, inputs_shape)
kernel = random.normal(rng2, (1, 16, 16))

inputs_flatten = inputs.reshape((8, 32, 32, 256))
kernel_flatten = kernel.reshape((1, 256))

@jax.jit
def dot_2d(inputs, kernel):
  return jax.lax.dot_general(inputs,  kernel,
                          (((3, 4), (1, 2)), ((), ())),
                          precision=jax.lax.Precision.HIGHEST)
@jax.jit
def dot_flatten(inputs, kernel):
  return jax.lax.dot_general(inputs,  kernel,
                          (((3,), (1,)), ((), ())),
                          precision=jax.lax.Precision.HIGHEST)

```

**Review by MostafaDehghani — APPROVED** [review]
